### ISSUE TITLE
max_workers is ambiguous

### ISSUE THREAD

**Issue by yadudoc** [Parsl/parsl#1429]
`max_workers` kwargs option for `HighThroughputExecutor` is ambiguous. Far too many users have asked what the option means, and I think we are better off renaming this as `max_workers_per_node`.

**Comment by benclifford** [comment]
Crossref #1353 

Maybe some phrasing subtlety when using fractional nodes: this parameter doesn't so directly limit the number of workers on a node in the case of two node-fractions being on the same name node.

**Comment by annawoodard** [comment]
(I think this is basically a dupe of #1353). As I mention there, I favor removing this option entirely. Otherwise, `max_workers_per_manager` is slightly more correct with respect to using fractional nodes.

**Comment by Andrew-S-Rosen** [comment]
I am reviving this old discussion because I just spent the last few hours in a Parsl death spiral because of a mismatch between the Parsl documentation's description of `max_workers` and its actual use. My hope is that this comment may help others. Tagging @WardLT because he runs calculations with this type of pattern a lot and surely is well-aware of this "gotcha" already.

The tl;dr --- to the best of my knowledge, `max_workers` is _not_ necessarily the number of workers on a node like [the documentation](https://parsl.readthedocs.io/en/stable/stubs/parsl.executors.HighThroughputExecutor.html) says. That is evident in the scenario below and likely is what @benclifford was alluding to (but I didn't discover this issue report until afterwards).

Here is the hypothetical scenario. We are launching PythonApps where each App is a Python script that calls some MPI-driven executable. The underlying MPI compute task uses a full node's worth of resources. We would like to run many of these PythonApps concurrently on a single Slurm allocation request. So, for instance, it would be a scenario where a single 10-node Slurm allocation would run, at most, 10 concurrent Apps where each App calls an executable that uses a full, dedicated node. For clarity, the Python script is calling `srun`, so everything is done using `SimpleLauncher()`.

If `max_workers` were indeed the number of workers per node, one would assume that it should be set to `max_workers=1` since we want each node to run a single, dedicated PythonApp at a time. However, doing this does not yield any concurrency. To achieve concurrency, one must actually set `max_workers` to be the number of concurrent Apps to run _even though we only want one App per node.

Here is what appears to be a proper configuration for the aforementioned scenario. Note how `max_workers` is not 1.

```python
import parsl
from parsl.config import Config
from parsl.executors import HighThroughputExecutor
from parsl.launchers import SimpleLauncher
from parsl.providers import SlurmProvider

account = "MyAccountName"

concurrent_jobs = 2
nodes_per_job = 1
cores_per_node = 128
min_slurm_allocations = 0
max_slurm_allocations = 1

config = Config(
    strategy="htex_auto_scale",
    executors=[
        HighThroughputExecutor(
            label="quacc_parsl",
            max_workers=concurrent_jobs,
            cores_per_worker=1e-6,
            provider=SlurmProvider(
                account=account,
                qos="debug",
                constraint="cpu",
                worker_init=f"module load vasp/6.4.1-cpu && export ASE_VASP_COMMAND='srun -N {nodes_per_job} --ntasks-per-node={cores_per_node} --cpu_bind=cores'",
                walltime="00:10:00",
                nodes_per_block=concurrent_jobs * nodes_per_job,
                init_blocks=0,
                min_blocks=min_slurm_allocations,
                max_blocks=max_slurm_allocations,
                launcher=SimpleLauncher(),
                cmd_timeout=60,
            ),
        )
    ],
)
```

This configuration will run only _one_ App at a time. In

**Comment by benclifford** [comment]
@Andrew-S-Rosen this sounds like something that should be dealt with as part of the htex MPI work for example in #3016. Without that, you're on your own for figuring out the parameters for awful hacks to run code that isn't running on a worker...

One comment I've made in that PR is that the MPI mode introduced in #3016 should be configured entirely differently - because of exactly the sort of thing you say above: notions like "workers" mean something completely different in MPI mode (as well as cores_per_worker, which otherwise people are generally configuring as "haha I hope this is 0, but not so 0 that it explodes" values)

See here onwards https://github.com/Parsl/parsl/pull/3016#issuecomment-1933017029.

Tagging @yadudoc for further discussion

**Comment by Andrew-S-Rosen** [comment]
Thanks for the very helpful comment, @benclifford! I'm glad this has already been receiving significant thought from the Parsl community. I completely agree with your comment about restructuring things. 

One question for @yadudoc. In our scenario [described above](https://github.com/Parsl/parsl/issues/1429#issuecomment-1937462128), we want Parsl to provision the appropriate resources and distribute the work appropriately but don't want Parsl to be writing any parallelization commands to input files (e.g. no `mpirun` or `srun` or anything). That's because the `PythonApp` is a Python script that makes `subprocess` calls of the MPI app where the call itself contains the `srun -N 1 --ntasks-per-node 64 ...` command. Is that also being addressed in #3016? I ask because I see `mpi_launcher="mpiexec"` in the docs and wasn't really sure how that would be relevant for my proposed use case.
